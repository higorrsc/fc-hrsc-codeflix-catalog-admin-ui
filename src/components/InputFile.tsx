import DeleteIcon from '@mui/icons-material/Delete';
import FileIcon from '@mui/icons-material/FileCopy';
import { IconButton, TextField } from '@mui/material';
import { useRef, useState } from 'react';

interface Props {
  onAdd: (files: FileList | null) => void;
  onRemove: (file: File) => void;
}

export const InputFile: React.FC<Props> = ({ onAdd, onRemove }) => {
  const [selectedFiles, setSelectedFiles] = useState<FileList | null>(null);

  const fileInputRef = useRef<HTMLInputElement>(null);

  const handleChange = (e: React.ChangeEvent<HTMLInputElement>) => {
    setSelectedFiles(e.target.files);
    onAdd(e.target.files);
  };

  const handleFileInput = () => {
    fileInputRef.current?.click();
  };

  const handleClear = () => {
    setSelectedFiles(null);
    if (selectedFiles) {
      Array.from(selectedFiles).forEach((file) => {
        onRemove(file);
      });
    }
  };

  return (
    <>
      <TextField
        type='text'
        placeholder='Select a file'
        value={selectedFiles?.length ? selectedFiles[0].name : ''}
        slotProps={{
          input: {
            readOnly: true,
            endAdornment: selectedFiles?.length ? (
              <IconButton onClick={handleClear}>
                <DeleteIcon />
              </IconButton>
            ) : (
              <IconButton onClick={handleFileInput}>
                <FileIcon />
              </IconButton>
            ),
          },
        }}
      />
      <input
        type='file'
        accept='*'
        name='inputFile'
        id='inputFile'
        style={{ display: 'none' }}
        ref={fileInputRef}
        onChange={handleChange}
      />
    </>
  );
};
