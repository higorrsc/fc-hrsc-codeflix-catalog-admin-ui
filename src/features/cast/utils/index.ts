import { CastMemberType } from '../types';

export function getCastMemberTypeOptions() {
  return Object.values(CastMemberType)
    .filter((value) => typeof value === 'number')
    .map((value) => ({
      value,
      label:
        CastMemberType[value as CastMemberType].charAt(0) +
        CastMemberType[value as CastMemberType].slice(1).toLowerCase(),
    }));
}
