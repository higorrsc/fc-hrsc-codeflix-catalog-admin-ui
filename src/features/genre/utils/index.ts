import { Category } from '../../category/types';
import { Genre } from '../types';

export const mapGenreToForm = (genre: Genre) => {
  return {
    id: genre.id,
    name: genre.name,
    is_active: genre?.is_active,
    categories_id: genre.categories?.map((category: Category) => category.id),
  };
};
