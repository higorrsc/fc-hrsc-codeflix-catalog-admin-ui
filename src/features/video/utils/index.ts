import { Video, VideoPayload } from 'src/features/video/types';

export const mapVideoToForm = (video: Video): VideoPayload => {
  return {
    id: video.id,
    title: video.title,
    description: video.description,
    year_launched: video.year_launched,
    opened: video.opened,
    rating: video.rating,
    duration: video.duration,
    genres_id: video.genres?.map((genre) => genre.id),
    categories_id: video.categories?.map((category) => category.id),
    cast_members_id: video.cast_members?.map((cast_member) => cast_member.id),
  };
};
