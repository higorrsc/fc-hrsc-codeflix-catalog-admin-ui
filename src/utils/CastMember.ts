import { CastMemberType } from '../types/CastMember';

const castMemberTypeLabelMap: Record<CastMemberType, string> = {
  [CastMemberType.ACTOR]: 'Actor',
  [CastMemberType.DIRECTOR]: 'Director',
};

export function getCastMemberTypeLabel(type: CastMemberType): string {
  return castMemberTypeLabelMap[type] || 'Unknown';
}

export const castMemberTypeOptions = [
  { value: CastMemberType.ACTOR, label: 'Actor' },
  { value: CastMemberType.DIRECTOR, label: 'Director' },
];
