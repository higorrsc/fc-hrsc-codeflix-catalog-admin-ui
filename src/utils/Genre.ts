import { Category } from '../types/Category';
import { Genre } from '../types/Genre';

export const mapGenreToForm = (genre: Genre) => {
  return {
    id: genre.id,
    name: genre.name,
    is_active: genre?.is_active,
    categories_id: genre.categories?.map((category: Category) => category.id),
  };
};
