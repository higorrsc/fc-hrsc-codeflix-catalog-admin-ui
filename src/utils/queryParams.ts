import { ListOptions } from '../types/_shared'; // Import ListOptions if needed for typing map

/**
 * Maps specific frontend parameter names (keys often matching ListOptions)
 * to their corresponding backend query parameter names expected by the API.
 * We use Partial<Record<keyof ListOptions | string, string>> to allow mapping
 * ListOptions keys and potentially other string keys if needed later.
 */
const paramKeyMap: Partial<Record<keyof ListOptions | string, string>> = {
  perPage: 'per_page',
  // Add other mappings if needed
};

/**
 * Parses an object containing potential query parameters into a URL query string
 * suitable for API requests.
 *
 * - Accepts an object like `ListOptions` or any object with potential query params.
 * - Filters out parameters with `null` or `undefined` values.
 * - Explicitly ignores UI-specific keys like `rowsPerPage`.
 * - Maps keys according to `paramKeyMap` (e.g., 'perPage' becomes 'per_page').
 * - Converts remaining parameter values to strings.
 *
 * @template T - The type of the params object (e.g., ListOptions).
 * @param params - An object containing potential query parameters.
 * @returns A URL query string (without the leading '?'). Returns an empty string
 *          if no valid parameters are found after filtering and mapping.
 */
export function parseQueryParams<T extends Record<string, any>>(
  params: T
): string {
  const queryParams = new URLSearchParams();

  Object.entries(params).forEach(([key, value]) => {
    // 1. Skip null/undefined/empty string for search
    if (
      value === null ||
      value === undefined ||
      (key === 'search' && value === '') // Also skip empty search strings
    ) {
      return;
    }

    // 2. Skip known UI-specific keys
    if (key === 'rowsPerPage') {
      return;
    }

    // 3. Determine the correct key name for the API
    const mappedKey = paramKeyMap[key as keyof typeof paramKeyMap] || key;

    // 4. Convert the value to a string and append
    queryParams.append(mappedKey, String(value));
  });

  return queryParams.toString();
}
