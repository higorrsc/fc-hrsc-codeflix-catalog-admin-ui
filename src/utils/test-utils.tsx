import type { RenderOptions } from '@testing-library/react';
import { render } from '@testing-library/react';
import React, { PropsWithChildren } from 'react';
import { Provider } from 'react-redux';

import { ThemeProvider } from '@mui/system';
import { configureStore } from '@reduxjs/toolkit';
import { SnackbarProvider } from 'notistack';
import { BrowserRouter } from 'react-router-dom';
import type { RootState } from '../app/store';
import { appTheme } from '../config/theme';
import { apiSlice } from '../features/api/apiSlice';
import { castMembersApiSlice } from '../features/cast/castMemberSlice';
import { categoriesApiSlice } from '../features/categories/categorySlice';
// As a basic setup, import your same slice reducers

// This type interface extends the default options for render from RTL, as well
// as allows the user to specify other things such as initialState, store.
interface ExtendedRenderOptions extends Omit<RenderOptions, 'queries'> {
  preloadedState?: Partial<RootState>;
  store?: ReturnType<typeof configureStore>;
}

export function renderWithProviders(
  ui: React.ReactElement,
  extendedRenderOptions: ExtendedRenderOptions = {}
) {
  const {
    // Automatically create a store instance if no store was passed in
    store = configureStore({
      reducer: {
        [categoriesApiSlice.reducerPath]: apiSlice.reducer,
        // @ts-ignore
        [castMembersApiSlice.reducerPath]: apiSlice.reducer,
      },
      // Adding the api middleware enables caching, invalidation, polling,
      // and other useful features of `rtk-query`.
      middleware: (getDefaultMiddleware) =>
        getDefaultMiddleware().concat(apiSlice.middleware),
    }),
    ...renderOptions
  } = extendedRenderOptions;

  const Wrapper = ({ children }: PropsWithChildren) => (
    <React.StrictMode>
      <BrowserRouter>
        <Provider store={store}>
          <ThemeProvider theme={appTheme}>
            <SnackbarProvider maxSnack={3}>{children}</SnackbarProvider>
          </ThemeProvider>
        </Provider>
      </BrowserRouter>
    </React.StrictMode>
  );

  // Return an object with the store and all of RTL's query functions
  return {
    store,
    ...render(ui, { wrapper: Wrapper, ...renderOptions }),
  };
}
