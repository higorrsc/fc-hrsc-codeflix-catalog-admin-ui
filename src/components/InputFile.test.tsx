import { fireEvent, render, screen } from '../utils/test-utils';
import { InputFile } from './InputFile';

describe('Test InputFile Component', () => {
  const mockOnAdd = jest.fn();
  const mockOnRemove = jest.fn();

  beforeEach(() => {
    mockOnAdd.mockClear();
    mockOnRemove.mockClear();
  });

  it('should render correctly initially', () => {
    const { asFragment } = render(
      <InputFile onAdd={mockOnAdd} onRemove={mockOnRemove} />
    );
    expect(screen.getByPlaceholderText('Select a file')).toBeInTheDocument();
    expect(screen.getByTestId('FileCopyIcon')).toBeInTheDocument();
    expect(screen.queryByTestId('DeleteIcon')).not.toBeInTheDocument();
    expect(asFragment()).toMatchSnapshot();
  });

  it('should call onAdd and update UI when a file is selected', () => {
    render(<InputFile onAdd={mockOnAdd} onRemove={mockOnRemove} />);

    const file = new File(['dummy content'], 'example.png', {
      type: 'image/png',
    });
    const inputElement = screen.getByTestId('input-file'); // Assuming you add a data-testid to the input

    // Simulate file selection
    fireEvent.change(inputElement, { target: { files: [file] } });

    expect(mockOnAdd).toHaveBeenCalledTimes(1);
    expect(mockOnAdd).toHaveBeenCalledWith(file);
    expect(screen.getByDisplayValue('example.png')).toBeInTheDocument();
    expect(screen.getByTestId('DeleteIcon')).toBeInTheDocument();
    expect(screen.queryByTestId('FileCopyIcon')).not.toBeInTheDocument();
  });

  it('should call onRemove and update UI when the delete button is clicked', () => {
    render(<InputFile onAdd={mockOnAdd} onRemove={mockOnRemove} />);

    const file = new File(['dummy content'], 'example.png', {
      type: 'image/png',
    });
    const inputElement = screen.getByTestId('input-file');

    // Select a file first
    fireEvent.change(inputElement, { target: { files: [file] } });

    expect(screen.getByDisplayValue('example.png')).toBeInTheDocument();
    const deleteButton = screen.getByTestId('DeleteIcon');

    // Click the delete button
    fireEvent.click(deleteButton);

    expect(mockOnRemove).toHaveBeenCalledTimes(1);
    expect(mockOnRemove).toHaveBeenCalledWith(file);
    expect(screen.getByPlaceholderText('Select a file')).toBeInTheDocument();
    expect(screen.queryByDisplayValue('example.png')).not.toBeInTheDocument();
    expect(screen.getByTestId('FileCopyIcon')).toBeInTheDocument();
    expect(screen.queryByTestId('DeleteIcon')).not.toBeInTheDocument();
  });

  it('should trigger file input click when FileIcon is clicked', async () => {
    render(<InputFile onAdd={mockOnAdd} onRemove={mockOnRemove} />);
    const fileIcon = screen.getByTestId('FileCopyIcon');
    const inputElement = screen.getByTestId('input-file') as HTMLInputElement;
    const inputClickSpy = jest.spyOn(inputElement, 'click');

    await fireEvent.click(fileIcon);

    expect(inputClickSpy).toHaveBeenCalledTimes(1);
    inputClickSpy.mockRestore();
  });

  it('should not call onAdd if no file is selected', () => {
    render(<InputFile onAdd={mockOnAdd} onRemove={mockOnRemove} />);
    const inputElement = screen.getByTestId('input-file');

    fireEvent.change(inputElement, { target: { files: [] } });
    expect(mockOnAdd).not.toHaveBeenCalled();

    fireEvent.change(inputElement, { target: { files: null } });
    expect(mockOnAdd).not.toHaveBeenCalled();
  });

  it('should not call onRemove if no file was previously selected when clear is called', () => {
    render(<InputFile onAdd={mockOnAdd} onRemove={mockOnRemove} />);
    expect(mockOnRemove).not.toHaveBeenCalled();
  });
});
