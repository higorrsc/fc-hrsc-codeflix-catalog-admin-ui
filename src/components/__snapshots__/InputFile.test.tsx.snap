// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test InputFile Component should render correctly initially 1`] = `
<DocumentFragment>
  <div
    class="MuiFormControl-root MuiTextField-root css-1xp5r68-MuiFormControl-root-MuiTextField-root"
  >
    <div
      class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-formControl MuiInputBase-adornedEnd Mui-readOnly MuiInputBase-readOnly css-jupps9-MuiInputBase-root-MuiOutlinedInput-root"
    >
      <input
        aria-invalid="false"
        class="MuiInputBase-input MuiOutlinedInput-input MuiInputBase-inputAdornedEnd Mui-readOnly MuiInputBase-readOnly css-1dune0f-MuiInputBase-input-MuiOutlinedInput-input"
        id="«r0»"
        placeholder="Select a file"
        readonly=""
        type="text"
        value=""
      />
      <button
        class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium css-53g0n7-MuiButtonBase-root-MuiIconButton-root"
        tabindex="0"
        type="button"
      >
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-1umw9bq-MuiSvgIcon-root"
          data-testid="FileCopyIcon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M16 1H4c-1.1 0-2 .9-2 2v14h2V3h12zm-1 4 6 6v10c0 1.1-.9 2-2 2H7.99C6.89 23 6 22.1 6 21l.01-14c0-1.1.89-2 1.99-2zm-1 7h5.5L14 6.5z"
          />
        </svg>
      </button>
      <fieldset
        aria-hidden="true"
        class="MuiOutlinedInput-notchedOutline css-1ll44ll-MuiOutlinedInput-notchedOutline"
      >
        <legend
          class="css-w4cd9x"
        >
          <span
            aria-hidden="true"
            class="notranslate"
          >
            ​
          </span>
        </legend>
      </fieldset>
    </div>
  </div>
  <input
    accept="*"
    data-testid="input-file"
    id="inputFile"
    name="inputFile"
    style="display: none;"
    type="file"
  />
</DocumentFragment>
`;
